src/librustc/mir/mod.rs

```diff
@@ -395,6 +411,10 @@ pub struct LocalDecl<'tcx> {
     /// True if this corresponds to a user-declared local variable.
     pub is_user_variable: bool,
 
+    /// True if this an internal local.
+    /// Such locals are not checked against the legal types in a generator.
+    pub internal: bool,
```

Comment:
If we're going in this direction - maybe it's better to just have the sanity check ignore private-address scalars because they can't contribute to soundness (someone could do a `impl !MyOibit for MyAlgebraicType`, but for a private-addr scalar you can bypass it by using a "fresh" k-value enum).